lib/app/migrate.ts

```diff
@@ -0,0 +1,584 @@
+import { ensureProjectExists } from '../projects';
+import { trackCommandMetadataUsage } from '../usageTracking';
+import SpinniesManager from '../ui/SpinniesManager';
+import { handleKeypress } from '../process';
+import {
+  checkMigrationStatus,
+  downloadProject,
+  migrateApp as migrateNonProjectApp_v2023_2,
+  beginMigration,
+  finishMigration,
+  listAppsForMigration,
+} from '@hubspot/local-dev-lib/api/projects';
+import { poll } from '../polling';
+import path from 'path';
+import { getCwd, sanitizeFileName } from '@hubspot/local-dev-lib/path';
+import { getHubSpotWebsiteOrigin } from '@hubspot/local-dev-lib/urls';
+import { extractZipArchive } from '@hubspot/local-dev-lib/archive';
+import { ArgumentsCamelCase } from 'yargs';
+import { MigrateAppOptions } from '../../types/Yargs';
+import chalk from 'chalk';
+import { validateUid } from '@hubspot/project-parsing-lib';
+import { MigrationApp } from '@hubspot/local-dev-lib/types/Project';
+import { UNMIGRATABLE_REASONS } from '@hubspot/local-dev-lib/constants/projects';
+import { mapToUserFacingType } from '@hubspot/project-parsing-lib/src/lib/transform';
+
+function getUnmigratableReason(reasonCode: string) {
+  switch (reasonCode) {
+    case UNMIGRATABLE_REASONS.UP_TO_DATE:
+      return i18n(
+        'commands.project.subcommands.migrateApp.migrationNotAllowedReasons.upToDate'
+      );
+    case UNMIGRATABLE_REASONS.IS_A_PRIVATE_APP:
+      return i18n(
+        'commands.project.subcommands.migrateApp.migrationNotAllowedReasons.isPrivateApp'
+      );
+    case UNMIGRATABLE_REASONS.LISTED_IN_MARKETPLACE:
+      return i18n(
+        'commands.project.subcommands.migrateApp.migrationNotAllowedReasons.listedInMarketplace'
+      );
+    default:
+      return i18n(
+        'commands.project.subcommands.migrateApp.migrationNotAllowedReasons.generic',
+        {
+          reasonCode,
+        }
+      );
+  }
+}
+
+async function handleMigrationSetup(
+  derivedAccountId: number,
+  options: ArgumentsCamelCase<MigrateAppOptions>
+) {
+  const { name, dest, appId } = options;
+  const { data } = await listAppsForMigration(derivedAccountId);
+
+  const { migratableApps, unmigratableApps } = data;
+  const allApps = [...migratableApps, ...unmigratableApps];
+
+  if (allApps.length === 0) {
+    throw new Error(
+      i18n(`commands.project.subcommands.migrateApp.errors.noApps`, {
+        accountId: derivedAccountId,
+      })
+    );
+  }
+
+  if (migratableApps.length === 0) {
+    const reasons = unmigratableApps.map(
+      app =>
+        `${chalk.bold(app.appName)}: ${getUnmigratableReason(app.unmigratableReason)}`
+    );
+
+    throw new Error(
+      `${i18n(`commands.project.subcommands.migrateApp.errors.noAppsEligible`, {
+        accountId: derivedAccountId,
+      })} \n  - ${reasons.join('\n  - ')}`
+    );
+  }
+
+  if (
+    appId &&
+    !allApps.some(app => {
+      return app.appId === appId;
+    })
+  ) {
+    throw new Error(
```

Comment:
Do we want to throw an error here, before the user chooses the app to migrate? Would that be confusing?

Reply:
This is checking if the user provided and appId that is not in the list of apps, so they have already chosen an app.